Core/include/Acts/TrackFinder/CKFSourceLinkSelector.hpp

```diff
@@ -0,0 +1,259 @@
+  /// @tparam source_link_t The type of source link
+  ///
+  /// @param calibrator The measurement calibrator
+  /// @param predictedParams The predicted track parameter on a surface
+  /// @param sourcelinks The pool of source links
+  ///
+  /// @return the compatible or outlier source link indice(s)
+  template <typename calibrator_t, typename source_link_t>
+  Result<std::pair<std::vector<size_t>, bool>> operator()(
+      const calibrator_t& calibrator, const BoundParameters& predictedParams,
+      const std::vector<source_link_t>& sourcelinks) const {
+    ACTS_VERBOSE("Invoked CKFSourceLinkSelector");
+
+    using CovMatrix_t = typename BoundParameters::CovMatrix_t;
+
+    auto surface = &predictedParams.referenceSurface();
+    // Get volume and layer ID
+    auto geoID = surface->geoID();
+    auto volumeID = geoID.volume();
+    auto layerID = geoID.layer();
+
+    // First check if the layer-level criteria is configured.
+    // If not, check if the volume-level criteria is configured.
+    // Otherwise, use world criteria
+    double chi2Cutoff = std::numeric_limits<double>::max();
+    size_t numSlsCutoff = std::numeric_limits<size_t>::max();
+    // Get the allowed maximum chi2 on this surface
+    while (true) {
+      // layer-level criteria
+      auto lvMaxChi2 = m_config.layerMaxChi2.find(volumeID);
+      if (lvMaxChi2 != m_config.layerMaxChi2.end()) {
+        auto lvlMaxChi2 = lvMaxChi2->second.find(layerID);
+        if (lvlMaxChi2 != lvMaxChi2->second.end()) {
+          chi2Cutoff = lvlMaxChi2->second;
+          break;
+        }
+      }
+      // volume-level criteria
+      auto vvMaxChi2 = m_config.volumeMaxChi2.find(volumeID);
+      if (vvMaxChi2 != m_config.volumeMaxChi2.end()) {
+        chi2Cutoff = vvMaxChi2->second;
+        break;
+      }
+      // world-level criteria
+      chi2Cutoff = m_config.maxChi2;
+      break;
+    }
+    ACTS_VERBOSE("Allowed maximum chisq: " << chi2Cutoff);
+
+    // Get the allowed maximum number of source link candidates on this surface
+    while (true) {
+      // layer-level criteria
+      auto lvMaxNumSls =
+          m_config.layerMaxNumSourcelinksOnSurface.find(volumeID);
+      if (lvMaxNumSls != m_config.layerMaxNumSourcelinksOnSurface.end()) {
+        auto lvlMaxNumSls = lvMaxNumSls->second.find(layerID);
+        if (lvlMaxNumSls != lvMaxNumSls->second.end()) {
+          numSlsCutoff = lvlMaxNumSls->second;
+          break;
+        }
+      }
+      // volume-level criteria
+      auto vvMaxNumSls =
+          m_config.volumeMaxNumSourcelinksOnSurface.find(volumeID);
+      if (vvMaxNumSls != m_config.volumeMaxNumSourcelinksOnSurface.end()) {
+        numSlsCutoff = vvMaxNumSls->second;
+        break;
+      }
+      // world-level criteria
+      numSlsCutoff = m_config.maxNumSourcelinksOnSurface;
+      break;
+    }
+    ACTS_VERBOSE("Allowed maximum number of source links: " << numSlsCutoff);
+
+    std::vector<std::pair<size_t, double>> candidateChi2;
```

Comment:
This makes sense. I have stored those vectors in the `CombinatorialKamanFilterResult`. This could avoid copy of those vectors when passing them between the source link selector and the `CombinatorialKamanFilter`.